src/model/reference.ml

```diff
@@ -170,43 +177,95 @@ let tokenize location s : token list =
         let kind, location = kind_ended identifier_location started_at index in
         scan_identifier index 0 (index - 1)
           ({ kind; identifier; location } :: tokens)
+    | '/' ->
+        let kind, location = kind_ended identifier_location started_at index in
+        scan_path index (index - 1) ({ kind; identifier; location } :: tokens)
     | _ ->
         scan_kind identifier identifier_location started_at (index - 1) tokens
   and kind_ended identifier_location started_at index =
     let offset = index + 1 in
     let length = started_at - offset in
-    let kind = Some (String.sub s offset length) in
+    let kind = `Prefixed (String.sub s offset length) in
     let location = Location_.in_string s ~offset ~length location in
     let location = Location_.span [ location; identifier_location ] in
     (kind, location)
+  and scan_path started_at index tokens =
+    (* The parsing rules are different for [/]-separated components. [-"".()] are
+       no longer meaningful. *)
+    match s.[index] with
+    | exception Invalid_argument _ -> path_ended started_at index :: tokens
+    | '/' -> scan_path index (index - 1) (path_ended started_at index :: tokens)
+    | _ -> scan_path started_at (index - 1) tokens
+  and path_ended started_at index =
+    let offset = index + 1 in
+    let length = started_at - offset in
+    let identifier = String.sub s offset length in
+    let location = Location_.in_string s ~offset ~length location in
+    { kind = `End_in_slash; identifier; location }
   in
 
   scan_identifier (String.length s) 0 (String.length s - 1) [] |> List.rev
 
-let expected allowed location =
-  let unqualified = "or an unqualified reference" in
+let expected ?(expect_paths = false) allowed location =
+  let unqualified = [ "an unqualified reference" ] in
+  let unqualified =
+    if expect_paths then "a path" :: unqualified else unqualified
+  in
   let allowed =
-    match allowed with
-    | [ one ] -> Printf.sprintf "'%s-' %s" one unqualified
-    | _ ->
-        String.concat ", "
-          (List.map (Printf.sprintf "'%s-'") allowed @ [ unqualified ])
+    String.concat ", " (List.map (Printf.sprintf "'%s-'") allowed @ unqualified)
   in
   expected_err allowed location
 
+(* Parse references that do not contain a [/]. Raises errors and warnings. *)
 let parse whole_reference_location s :
     Paths.Reference.t Error.with_errors_and_warnings =
   let open Paths.Reference in
   let open Names in
+  let rec path components next_token tokens : Hierarchy.t =
+    match (next_token.kind, tokens) with
```

Comment:
No because that adds the `[]` case, which a lot of effort was made to avoid it. Also, unpacking the kind from that is too much syntax overhead.

Reply:
I think that then `match next_token, tokens with ...` is a little bit more clear. I find those:
```
| `End_in_slash, [] when next_token.identifier = "." ->
```
hard to read, since it's hard to remember that `` `End_in_slash `` and `next_token` are linked at this point 
Those cases amounts to 3 out of 6, but we are likely to add more of then.

```
| {kind = `End_in_slash ; identifier = "" ; _ }, [] ->
```
Feels easier to read to me.

(I will obviously not die on this hill, if you are not convinced just resolve the conversation!)

Reply:
Good idea, I implemented that.